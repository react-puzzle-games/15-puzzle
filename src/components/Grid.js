import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Tile, { propTypes as TilePropTypes } from './Tile';

class Grid extends Component {
  render() {
    const {
      className,
      tiles,
      onTileClick,
      gridSize,
    } = this.props;

    return (
      <div className={className}>
        <div className="tiles">
          {tiles.map((tile, tileId) => {
            return (
              <Tile
                {...tile}
                key={`tile-${tileId}`}
                correct={tile.tileId + 1 === tile.number}
                onClick={onTileClick}
                visible={tile.number < gridSize ** 2}
              />
            );
          })}
        </div>
      </div>
    );
  }
}

Grid.propTypes = {
  tiles: PropTypes.arrayOf(PropTypes.shape(TilePropTypes)).isRequired,
  gridSize: PropTypes.number.isRequired,
  tileSize: PropTypes.number.isRequired,
  onTileClick: PropTypes.func.isRequired,
};

export default styled(Grid)`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 500px;

  & .tiles {
    width: ${props => props.tileSize * props.gridSize}px;
    height: ${props => props.tileSize * props.gridSize}px;
    position: relative;
    text-align: center;
  }
`;
