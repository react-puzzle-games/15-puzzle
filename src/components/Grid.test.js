import renderer from 'react-test-renderer';
import React from 'react';
import Grid from './Grid';

describe('Grid renders correctly', () => {
  test('given an empty set of tiles', () => {
    const tree = renderer
      .create(
        <Grid tiles={[]} onTileClick={() => {}} tileSize={90} gridSize={4} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('given one tile', () => {
    const tree = renderer
      .create(
        <Grid
          tiles={[
            {
              tileId: 3,
              width: 90,
              height: 90,
              left: 0,
              top: -180,
            },
          ]}
          onTileClick={() => {}}
          tileSize={90}
          gridSize={1}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('given a set of tiles', () => {
    const tree = renderer
      .create(
        <Grid
          tiles={[
            {
              tileId: 0,
              width: 90,
              height: 90,
              left: 0,
              top: 0,
              number: 4,
            },
            {
              tileId: 1,
              width: 90,
              height: 90,
              left: 90,
              top: 0,
              number: 2,
            },
            {
              tileId: 2,
              width: 90,
              height: 90,
              left: 0,
              top: 90,
              number: 3,
            },
            {
              tileId: 3,
              width: 90,
              height: 90,
              left: 90,
              top: 90,
              number: 1,
            },
          ]}
          onTileClick={() => {}}
          tileSize={90}
          gridSize={1}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
