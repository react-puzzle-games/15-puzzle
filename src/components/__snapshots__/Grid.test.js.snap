// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly given a set of tiles 1`] = `
<div
  className="sc-bwzfXH jBBIHp"
>
  <div
    className="tiles"
  >
    <div
      className="sc-bdVaJa jodVcS tile-invisible"
      onClick={[Function]}
    >
      <span
        className="tile-number"
      >
        4
      </span>
    </div>
    <div
      className="sc-bdVaJa gaQDFA tile-invisible"
      onClick={[Function]}
    >
      <span
        className="tile-number"
      >
        2
      </span>
    </div>
    <div
      className="sc-bdVaJa htVqMf tile-invisible"
      onClick={[Function]}
    >
      <span
        className="tile-number"
      >
        3
      </span>
    </div>
    <div
      className="sc-bdVaJa dQlHZv tile-invisible"
      onClick={[Function]}
    >
      <span
        className="tile-number"
      >
        1
      </span>
    </div>
  </div>
</div>
`;

exports[`Grid renders correctly given an empty set of tiles 1`] = `
<div
  className="sc-bwzfXH iPaWuQ"
>
  <div
    className="tiles"
  />
</div>
`;

exports[`Grid renders correctly given one tile 1`] = `
<div
  className="sc-bwzfXH jBBIHp"
>
  <div
    className="tiles"
  >
    <div
      className="sc-bdVaJa eHNnrm tile-invisible"
      onClick={[Function]}
    >
      <span
        className="tile-number"
      >
        0
      </span>
    </div>
  </div>
</div>
`;
